website/server/libs/groupTasks.js

```diff
@@ -1,24 +1,68 @@
 import * as Tasks from '../models/task';
+import {model as Groups} from '../models/group';
+import {model as Users} from '../models/user/index';
 
 const SHARED_COMPLETION = {
   default: 'recurringCompletion',
   single: 'singleCompletion',
   every: 'allAssignedCompletion',
 };
 
-async function _completeMasterTask (masterTask) {
-  masterTask.completed = true;
+// @TODO Group task scoring
+async function _completeOrUncompleteMasterTask (masterTask, completed) {
+  masterTask.completed = completed;
   await masterTask.save();
 }
 
-async function _deleteUnfinishedTasks (groupMemberTask) {
-  await Tasks.Task.deleteMany({
-    'group.taskId': groupMemberTask.group.taskId,
-    $and: [
-      {userId: {$exists: true}},
-      {userId: {$ne: groupMemberTask.userId}},
-    ],
-  }).exec();
+async function _updateAssignedUsersTasks (masterTask, groupMemberTask) {
+  if (groupMemberTask.type === 'todo') {
+    if (groupMemberTask.completed) {
+      // The task was done by one person and is removed from others' lists
+      await Tasks.Task.deleteMany({
+        'group.taskId': groupMemberTask.group.taskId,
+        $and: [
+          {userId: {$exists: true}},
+          {userId: {$ne: groupMemberTask.userId}},
+        ],
+      }).exec();
+    } else {
+      // The task was uncompleted by the group member and should be recreated for assignedUsers
+      let group = await Groups.findById(masterTask.group.id).exec();
+      let userList = [];
+      masterTask.group.assignedUsers.forEach(userId => {
+        let query = {_id: userId};
+        userList.push(query);
+      });
+      await Users.find({
+        $or: userList,
+      }).then(async assignedUsers => {
+        for (const assignedUser of assignedUsers) {
+          let promises = [];
+          promises.push(group.syncTask(masterTask, assignedUser));
+          promises.push(group.save());
+          await Promise.all(promises);
+        }
+      });
+    }
+  } else {
+    // Complete or uncomplete the task on other users' lists
+    await Tasks.Task.find({
+      'group.taskId': groupMemberTask.group.taskId,
+      $and: [
+        {userId: {$exists: true}},
+        {userId: {$ne: groupMemberTask.userId}}
+      ]}).then(tasks => {
```

Comment:
why is then used here? Instead of something like `const tasks = await Tasks.Task.find(..) ` and then iterate on them?

Reply:
It seemed an appropriate pattern, though I do not have a lot of experience here.

Why not use `then`, and what's better about splitting it out like that?

Reply:
Sorry, I didn't explain myself well.

Using `.then` with promises isn't error prone at all but it's the same as doing `await promise` and when you start using async/await or then/catch it's better to only use one method only to avoid confusion. We prefer async/await as it's usually easier to follow and read the code

Reply:
Ah, this makes complete sense.